### PR TITLE
Create transparency with respect to the metric’s data and model applicability

### PR DESCRIPTION
**Pull request by annahedstroem**
### Addressed issues:
- https://github.com/understandable-machine-intelligence-lab/Quantus/issues/280
- https://github.com/understandable-machine-intelligence-lab/Quantus/issues/202
- Renaming from `last_results` to `evaluation_scores` and `all_results` to  `all_evaluation_scores`

### Minimum acceptance criteria
- Specify what is necessary for the PR to be merged with the main branch.
- @mentions of the person that is apt to review these changes e.g., @annahedstroem

### CODE REVIEW

**Comment by aaarrti** [comment]
> ### Addressed issues:
> * [Create transparency with respect to the metric’s individual data and model applicability #280](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/280)
> * [Normalisation per category #202](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/202)
> 
> ### Minimum acceptance criteria
> * Specify what is necessary for the PR to be merged with the main branch.
> * @mentions of the person that is apt to review these changes e.g., @annahedstroem

Just to make sure we are on the same page, the goals of this task are:
- make it easier for users to understand if their data and metric of choice are compatible 
- systematize available metrics, pre-processing features, supported model, etc.

Am I right?

File: quantus/metrics/base.py [hunk 1]
```diff
@@ -101,6 +116,37 @@ def __init__(
...
+            or not hasattr(self, "_data_applicability")
+            or not hasattr(self, "_model_applicability")
+            or not hasattr(self, "_score_direction")
+            or not hasattr(self, "_evaluation_category")
+        ):
+            raise NotImplementedError(
+                "Subclasses must define name, data_applicability, model_applicability and score_direction before instantiation"
+            )
+
+    @property
```
Comment (aaarrti) [inline quantus/metrics/base.py:130]: Why not just
```python
@property
@abstractmethod
def name(self):
     raise NotImplementedError
```
This way we won't need additional "private" properties, and will enforce inheritors to specify a name.

File: quantus/metrics/base.py [hunk 2]
```diff
@@ -18,13 +18,28 @@
...
 from quantus.helpers.model.model_interface import ModelInterface
+from quantus.functions import postprocess_func
+from quantus.helpers.enums import ModelType, DataType, ScoreDirection, EvaluationCategory, EvaluationCategory
 
 
 class Metric:
     """
     Implementation of the base Metric class.
+
+    Attributes:
```
Comment (aaarrti) [inline quantus/metrics/base.py:29]: Typically, attributes, which start with underscore, are supposed to be private, 
meaning, they should not part of the public API, and typically are left out of doc strings. Are you sure they should be mentioned in docstring?

File: quantus/helpers/enums.py [hunk 3]
```diff
@@ -0,0 +1,61 @@
...
+class DataType(Enum):
+    """
+    This enum represents the different types of data that a metric implementation currently supports.
+
+        - IMAGE: Represents image data.
+        - TABULAR: Represents tabular data.
+        - TEXT: Represents text data.
+    """
+
+    IMAGE = "image"  # 3D data
```
Comment (aaarrti) [inline quantus/helpers/enums.py:13]: `IMAGE`, `TIME-SERIES` and `TABULAR` describe the shapes of the tensors.
While `TEXT` the underlying data type.

Those two look like separate, independent categories to me.
Ideally, we should also specify which data types we expect (or support) `float32`, `float64` or something else?

And finally, for images it would be really useful to make it visible if we expect 
pixel values in the range [0, 255] or [0, 1]. This one always leaves me puzzled 🙃

File: quantus/helpers/enums.py [hunk 4]
```diff
@@ -0,0 +1,61 @@
+from enum import Enum
```
Comment (aaarrti) [inline quantus/helpers/enums.py:1]: I guess, this is more a question of test. What do you think of using `typing.Literal`? E.g., 

```
from typing import Literal 

DataType = Literal["text", "image", "time-series"]
```

No real benefit, though 🤷‍♂️

File: quantus/helpers/enums.py [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+        - TEXT: Represents text data.
+    """
+
+    IMAGE = "image"  # 3D data
+    TIMESERIES = "time-series"  # 1D data
+    TABULAR = "tabular"  # 2D data
+    TEXT = "text"
+
+
+class ModelType(Enum):
```
Comment (aaarrti) [inline quantus/helpers/enums.py:19]: I'd say `torch` and `tensorflow` describe the underlying framework used for tensor computation, and `ModelType` sound like it is about architecture or task it is solving, so this one is a bit confusing. It would also be great to have some sort of list of supported DNN libraries, since they all have slightly different API, e.g., `torchvision`, `keras`, `transformers`, `keras_cv` ...

File: quantus/metrics/base.py [hunk 6]
```diff
@@ -101,6 +116,37 @@ def __init__(
...
+    def name(self):
+        return self._name
+
+    @property
+    def data_applicability(self):
+        return self._data_applicability
+
+    @property
+    def model_applicability(self):
+        return self._model_applicability
```
Comment (aaarrti) [inline quantus/metrics/base.py:140]: I think all metrics support `torch` and `tensorflow`. 
Maybe we could return here both as default, and just in case 1 metric doesn't
we can override the property there.

File: quantus/metrics/robustness/relative_representation_stability.py [hunk 7]
```diff
@@ -22,21 +22,37 @@
...
 from quantus.functions.perturb_func import uniform_noise, perturb_batch
 from quantus.helpers.utils import expand_attribution_channel
+from quantus.helpers.enums import ModelType, DataType, ScoreDirection, EvaluationCategory
 
 
 class RelativeRepresentationStability(BatchedPerturbationMetric):
     """
-    Relative Output Stability leverages the stability of an explanation with respect
-    to the change in the output logits
+    Relative Representation Stability leverages the stability of an explanation with respect
```
Comment (aaarrti) [inline quantus/metrics/robustness/relative_representation_stability.py:30]: Oh, that's a typo I did 🤦 
Thanks for noticing!

File: quantus/metrics/base.py [hunk 8]
```diff
@@ -228,6 +274,8 @@ def __call__(
         # Call custom post-processing.
         self.custom_postprocess(**data)
 
+        self.skill_scores = self.compute_skill_score(a_batch=data["a_batch"])
```
Comment (aaarrti) [inline quantus/metrics/base.py:277]: Where is `compute_skill_score` method defined?

File: quantus/metrics/robustness/relative_representation_stability.py (hunk 7 above)
Comment (annahedstroem) [reply quantus/metrics/robustness/relative_representation_stability.py:30]: np!

File: quantus/helpers/enums.py (hunk 5 above)
Comment (annahedstroem) [reply quantus/helpers/enums.py:19]: yeah I was thinking that this one could be better developed in the future as we expand the modeltypes (ModelInterface.NewModel) and the list of ModelTypes will basically be those

File: quantus/helpers/enums.py (hunk 4 above)
Comment (annahedstroem) [reply quantus/helpers/enums.py:1]: hmm! Literal was introduced in Python3.8, so for our package, enum might be better?

File: quantus/helpers/enums.py (hunk 3 above)
Comment (annahedstroem) [reply quantus/helpers/enums.py:13]: > IMAGE, TIME-SERIES and TABULAR describe the shapes of the tensors. While TEXT the underlying data type.

I agree, the text one is vague. I kept it there more as a placeholder until we further define it with an actual usecase.

> Those two look like separate, independent categories to me.

Could be so, should we develop it further when we have the first text model wrapped into Quantus?

> Ideally, we should also specify which data types we expect (or support) float32, float64 or something else?

We should definitely get more precise. Should we define this in a separate PR? I would say that this PR does a first good basic job and then we can go more in detail in a follow-up PR.

> And finally, for images it would be really useful to make it visible if we expect pixel values in the range [0, 255] or [0, 1]. This one always leaves me puzzled 🙃

Hmmm. I think this is out of the scope of Quantus since it is related to the model training process..... e..g, whether the model is trained on [0,1] and not [0, 255] then, ultimately this is also what should be passed as input on test (explanation/ evaluation) time.

File: quantus/metrics/base.py (hunk 2 above)
Comment (annahedstroem) [reply quantus/metrics/base.py:28]: Hmm good point. But should we instead make them public directly? Why private actually...... this is useful information for the users.

Comment (annahedstroem) [reply quantus/metrics/base.py:28]: We can just skip the whole `@property def get_name`  ... by stating them as public attributes straight away.

File: quantus/metrics/base.py (hunk 1 above)
Comment (annahedstroem) [reply quantus/metrics/base.py:131]: Good idea, will implement.

**Comment by annahedstroem** [comment]
> > ### Addressed issues:
> > 
> > * [Create transparency with respect to the metric’s individual data and model applicability #280](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/280)
> > * [Normalisation per category #202](https://github.com/understandable-machine-intelligence-lab/Quantus/issues/202)
> > 
> > ### Minimum acceptance criteria
> > 
> > * Specify what is necessary for the PR to be merged with the main branch.
> > * @mentions of the person that is apt to review these changes e.g., @annahedstroem
> 
> Just to make sure we are on the same page, the goals of this task are:
> 
> * make it easier for users to understand if their data and metric of choice are compatible
> * systematize available metrics, pre-processing features, supported model, etc.
> 
> Am I right?

Yes, this is a starting point for this effort. Then we can complement this PR with a more in-depth ModelType and DataType checker, as the need arise.

File: quantus/helpers/enums.py (hunk 4 above)
Comment (aaarrti) [reply quantus/helpers/enums.py:1]: then the question is closed 🤷

File: quantus/helpers/enums.py (hunk 3 above)
Comment (aaarrti) [reply quantus/helpers/enums.py:13]: Sounds good, I will do a proposal PR with shapes and `dtypes` differentiation on top of this one in the next time then

File: quantus/metrics/base.py (hunk 2 above)
Comment (aaarrti) [reply quantus/metrics/base.py:43]: I'd say completely remove them, since they don't really make sense for base class. Moreover, because of the same name, the descriptors are already defined on the class level, which renders 
```python
@property
@abstractmethod
```
useless. I like the idea of abstract properties much more, since they will enforce each metric to implement them.

File: quantus/metrics/base.py [hunk 9]
```diff
@@ -98,8 +129,28 @@ def __init__(
...
+    @abstractmethod
+    def name(self):
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def evaluation_category(self):
+        raise NotImplementedError
+
+    @property
```
Comment (aaarrti) [inline quantus/metrics/base.py:145]: This one can safely be implemented in the base class, 
```python
@property
def model_applicability(self) -> Set[ModelType]:
     return {ModelType.Torch, ModelType.TensorFlow}
```
Since all metrics support both, afaik.

File: quantus/metrics/base.py [hunk 10]
```diff
@@ -98,8 +129,28 @@ def __init__(
...
+    @abstractmethod
+    def model_applicability(self):
+        raise NotImplementedError
+
+    @property
+    @abstractmethod
+    def data_applicability(self):
```
Comment (aaarrti) [inline quantus/metrics/base.py:152]: Please add type hints for properties.

File: quantus/metrics/base.py (hunk 6 above)
Comment (annahedstroem) [reply quantus/metrics/base.py]: for consistency, let's type everything

Comment (annahedstroem) [reply quantus/metrics/base.py]: especially in the future, we might have less metrics coverage for certain metrics

File: quantus/metrics/base.py (hunk 9 above)
Comment (annahedstroem) [reply quantus/metrics/base.py:145]: Sure, can be done.

File: quantus/metrics/base.py (hunk 2 above)
Comment (annahedstroem) [reply quantus/metrics/base.py]: Should be fixed now.
